source/vir/src/interpreter.rs

```diff
@@ -0,0 +1,1391 @@
+                                        int_i,
+                                    ));
+                                    let args = Arc::new(vec![boxed_i]);
+                                    let call = exp_new(CallLambda(
+                                        lambda.typ.clone(),
+                                        lambda.clone(),
+                                        args,
+                                    ));
+                                    eval_expr_internal(ctx, state, &call)
+                                })
+                                .collect();
+                            let im_vec: Vector<Exp> = vec?.into_iter().collect();
+                            seq_new(im_vec)
+                        }
+                        _ => ok,
+                    }
+                }
+                Push => match &args[0].x {
+                    Interp(Seq(s)) => {
+                        let mut s = s.clone();
+                        s.push_back(args[1].clone());
+                        seq_new(s)
+                    }
+                    _ => ok,
+                },
+                Update => match &args[0].x {
+                    Interp(Seq(s)) => match get_int(&args[1]) {
+                        Some(index) if index < s.len() => {
+                            let s = s.update(index, args[2].clone());
+                            seq_new(s)
+                        }
+                        _ => ok,
+                    },
+                    _ => ok,
+                },
+                Subrange => match &args[0].x {
+                    Interp(Seq(s)) => {
+                        let start = get_int(&args[1]);
+                        let end = get_int(&args[2]);
+                        match (start, end) {
+                            (Some(start), Some(end)) if start <= end && end <= s.len() => {
+                                seq_new(s.clone().slice(start..end))
+                            }
+                            _ => ok,
+                        }
+                    }
+                    _ => ok,
+                },
+                Add => match (&args[0].x, &args[1].x) {
+                    (Interp(Seq(s1)), Interp(Seq(s2))) => {
+                        let mut s = s1.clone();
+                        s.append(s2.clone());
+                        seq_new(s)
+                    }
+                    _ => ok,
+                },
+                Len => match &args[0].x {
+                    Interp(Seq(s)) => int_new(BigInt::from_usize(s.len()).unwrap()),
+                    _ => ok,
+                },
+                Index => match &args[0].x {
+                    Interp(Seq(s)) => match &args[1].x {
+                        UnaryOpr(crate::ast::UnaryOpr::Box(_), e) => match &e.x {
+                            Const(Constant::Int(index)) => {
+                                let index = BigInt::to_usize(index).unwrap();
```

Comment:
> this can't fail because the value is negative (type checker requires it to be nat) 

The signature for `seq.index` is: 
```
pub spec fn index(self, i: int) -> A
```
which allows a negative value, right?  So we would need to handle that too, no?

> Can you handle the case instead of panicking?

What would be the right way to handle it?  Since we're representing the sequences using a Vector, we need a `usize` index.  Or are you suggesting I just pick an arbitrary integer (e.g., 42) in the case where the provided index is too large to convert (and possibly issue a warning)?

Reply:
Whoops, you're right. Still the point stands, we should handle the case rather than panic.

I guess it would technically be sound to pick an arbitrary integer (in the sense that we've left enough things unconstrained until now) but I think that behavior would be odd and unexpected.

 You could just leave the expression unsimplified and issue a warning, like you'd otherwise do when you can't simplify, right?

Reply:
> You could just leave the expression unsimplified and issue a warning, like you'd otherwise do when you can't simplify, right?

Yes, that's a good point.  Done.